core/src/main/scala/com/evolutiongaming/catshelper/ReadWriteRef.scala

```diff
@@ -0,0 +1,160 @@
+
+import cats.Applicative
+import cats.effect.{Concurrent, Resource}
+import cats.effect.concurrent.Ref
+import cats.implicits._
+
+/**
+ * A mutable reference to `A` value with read-write lock semantics:
+ *  - multiple "read" operations are allowed as long as there's no "write" running.
+ *  - "write" operation is exclusive.
+ *  - NB: "read" can't be promoted to "write" yet. Attempting a "write" inside "read" will deadlock.
+ *
+ * Build it with [[ReadWriteRef.PartialApply.of `ReadWriteRef[F].of(a)`]] or
+ * [[ReadWriteRef.of `ReadWriteRef.of[F, A](a)`]].
+ *
+ * @note Both [[ReadWriteRef.read `read`]] and [[ReadWriteRef.write `write`]] are implemented as `Resource`
+ *       since it has clear mental model of something that may be "in use".
+ */
+trait ReadWriteRef[F[_], A] {
+
+  /**
+   * "Read" operation. Multiple reads may be in use simultaneously.
+   * `reads.use(…)` blocks semantically while [[write]] in in use.
+   */
+  def read: Resource[F, A]
+
+  /**
+   * "Write" operation. Requires exclusive access.
+   * `write.use(…)` blocks semantically while [[read]] or another [[write]] is in use.
+   */
+  def write: Resource[F, ReadWriteRef.Upd[F, A]]
+}
+
+object ReadWriteRef {
+  /**
+   * A "resource" of [[ReadWriteRef.write `write`]].
+   * `(A => F[A])` part lets you read current value and produce a new one, effectfully.
+   * `F[A]` in the return part lets you use new `A` in the outer context, e.g. return it from `use`.
+   */
+  trait Upd[F[_], A] extends ((A => F[A]) => F[A])
+
+  object Upd {
+    implicit final class SelfOps[F[_], A](val self: Upd[F, A]) extends AnyVal {
+      /** A shorthand for non-effectful updates. */
+      def plain(f: A => A)(implicit F: Applicative[F]): F[A] = self(a => F.pure(f(a)))
+
+      /** A shorthand for non-effectful update that ignores previous value. */
+      def const(a: A)(implicit F: Applicative[F]): F[A] = self(_ => F.pure(a))
+    }
+  }
+
+  final class PartialApply[F[_]: Concurrent] {
+    def of[A](init: A): F[ReadWriteRef[F, A]] = ReadWriteRef.of[F, A](init)
+  }
+
+  def apply[F[_]: Concurrent]: PartialApply[F] = new PartialApply[F]
+
+  def of[F[_], A](init: A)(implicit F: Concurrent[F]): F[ReadWriteRef[F, A]] = {
+    case class State(
+      a: A,
+      inUse: Int = 0, // -1 -- exclusive write, 0 -- not in use, >0 -- readers
+      pendingReads: List[A => F[Unit]] = List.empty,
+      pendingWrites: List[A => F[Unit]] = List.empty,
+    )
+
+    Ref[F].of(State(init)) map { stateRef =>
+      // The `release` function "commits" A back to the state, adjusts `inUse` count,
+      // and notifies pending listeners.
+      //
+      // IMPORTANT:
+      //   This function must be called only from uncancellable blocks (e.g. Resource release).
+      def release(a: A, useChange: Int => Int): F[Unit] = stateRef
+        .modify { s0 =>
+          val inUse = useChange(s0.inUse)
+
+          s0.pendingWrites match {
+            // Give writers a priority
```

Comment:
why do we change the order?

I think calls should be executed in the same order as originally made and not just we do writes first and there after proceed with reads

Reply:
This is not about changing the order. It's more about throughput. Readers execute all-at-once, writers – one-by-one. They have separate "queues" for that exact reason.

My idea of pushing writers forward was
- to let readers have more up-to-date value
- letting readers pile-up and then getting rid of them in one go

I can imagine that in pathological cases writers can overwhelm readers, but perhaps we can optimise for that case a bit later.

Reply:
I think this might be very confusing in some cases, when execution happens in other than call order.

Reply:
But `ReadWriteRef` is all about re-ordering in a sense that writers have to wait while reads are running, and more reads may get ahead of them. That's the difference between this one and `SerialRef`.

Reply:
the case I'm talking about is for instance you made a `write` and there after `read`, and you may end up in situation when `read` overtook `write`

Reply:
"After" implies a monadic sequence in which case `read` won't even be attempted before `write`. And without such sequencing you have a concurrent execution where anything can happen ;)

Reply:
Buuut, speaking of concurrent execution and ordering, let's focus on another piect of code which we may want to improve.

Right now we execute pending reads like this
```scala
val effect = s0.pendingReads.traverse_(l => l(a))
```
And `pendingReads` is a `List` which is built by _prepending_ late readers…
So we have reordering and _sequential_ execution. Should we improve both? If yes – how?

Reply:
The most obvious would be to reverse this list (or use a `Queue`) and do `parTraverse_`. Do you see any drawbacks here?

Reply:
> "After" implies a monadic sequence in which case read won't even be attempted before write. And without such sequencing you have a concurrent execution where anything can happen ;)

```scala

val promise: ...
read.use { promise.get }.toFuture

write.use.toFuture
read.use.toFuture

promise.complete()
```

I know, that's an extreme case :)

Reply:
Sure, but that exactly shows how `ReadWriteRef` works – the second `read.use(…).toFuture` will execute in parallel with the first read and has all the chances to finish first.

Reply:
I would do stuff like change a state a bit - so you have `state:List[Action]` where
```
sealed trait Action
object Action {
 case class Reads(value: List[A => F[Unit]])
 case class Write(value: A => F[Unit])
```

So this way you collect reads in between writes and execute them in `par` (you can do that with Parallel or still remain with Concurrent)

Reply:
writes are executed strictly serially

Reply:
regardless of this discussion I can accept this PR. Just wanted to make a point here

Reply:
Hmm this would change the execution model considerably. Right now readers can execute immediately as long as there is no write running.

If I get you right, you suggest blocking readers as soon as there is a writer _pending_. It's not wrong, I can even agree that it's more fair in its strictness. Okay, let me think of it a bit.

Reply:
I even think on renaming a bit for clarity 

```
sealed trait Serial
object Serial {
 case class Parallel(reads: List[A => F[Unit]]) extends Serial
 case class Write(value: A => F[Unit]) extends Serial
```

Reply:
Reimplemented in acf2ba6e.